### PR TITLE
Fluxework

### PR DESCRIPTION
**Pull request by geralddejong**
Many UI changes, fundamental bug fixes, removal of global clock, etc.

### CODE REVIEW

File: client/.eslintrc.js [hunk 1]
```diff
@@ -24,7 +24,7 @@ module.exports = {
         },
       },
     ],
-    "@typescript-eslint/indent": ["error", 4],
+    "@typescript-eslint/indent": "off",
```
Comment (mvanderh) [inline client/.eslintrc.js]: It was complaining too much?

File: client/src/fabric/fabric-kernel.ts [hunk 2]
```diff
@@ -215,8 +215,9 @@ export class FabricInstance {
         return this.fabricIndex
     }
 
-    public recycle(): void {
-        this.recycleFabric(this.fabricIndex)
+    public release(): void {
+        this.releaseInstance(this.fabricIndex)
+        this.ex.reset()
```
Comment (mvanderh) [inline client/src/fabric/fabric-kernel.ts]: Still would prefer `ex` to be renamed to `engine` or something

File: client/src/fabric/physics.ts [hunk 3]
```diff
@@ -110,7 +115,9 @@ export class Physics {
             },
             factor$: factor,
             defaultValue: isGlobal ? 1.0 : factor.getValue(),
+            apply: (instance: FabricInstance) => applyFeature(instance, feature),
         }
+        return feature
```
Comment (mvanderh) [inline client/src/fabric/physics.ts]: I would inline `feature` here. Less code = better

File: client/src/fabric/tensegrity-brick.ts [hunk 4]
```diff
@@ -100,10 +102,12 @@ export function createBrickOnFace(face: IFace): IBrick {
...
     const y = new Vector3().crossVectors(z, x).normalize()
-    const xform= new Matrix4().makeBasis(x, y, z).setPosition(midpoint)
-    const points = createBrickPointsOnOrigin(Triangle.PPP, 0.8, 1.0)
+    const xform = new Matrix4().makeBasis(x, y, z).setPosition(midpoint)
+    const base = negativeFace ? Triangle.NNN : Triangle.PPP
+    const points = createBrickPointsOnOrigin(base, 0.8, 1.0)
     const movedToFace = points.map(p => p.applyMatrix4(xform))
-    return createBrick(brick.fabric, movedToFace, Triangle.NNN)
+    const baseTriangle = negativeFace ? Triangle.PPP : Triangle.NNN
+    return createBrick(brick.fabric, movedToFace, baseTriangle)
```
Comment (mvanderh) [inline client/src/fabric/tensegrity-brick.ts]: Still voodoo to me :)

File: client/src/fabric/tensegrity-brick.ts [hunk 5]
```diff
@@ -90,6 +91,7 @@ export function createBrickOnOrigin(fabric: TensegrityFabric, altitude: number):
 }
 
 export function createBrickOnFace(face: IFace): IBrick {
+    const negativeFace = TRIANGLE_ARRAY[face.triangle].negative
```
Comment (mvanderh) [inline client/src/fabric/tensegrity-brick.ts]: Has to be a better name for `TRIANGLE_ARRAY`

File: client/src/gotchi/gotchi.ts [hunk 6]
```diff
@@ -30,7 +31,8 @@ export class Gotchi {
         private genome: Genome,
         private gotchiFactory: IGotchiFactory,
     ) {
-        if (body.isGestating) {
+        const bodyIsBusy = true
+        if (bodyIsBusy) {
```
Comment (mvanderh) [inline client/src/gotchi/gotchi.ts]: Lol

Comment (mvanderh) [reply client/src/gotchi/gotchi.ts]: Might wanna add a TODO comment here

File: client/src/index.tsx [hunk 7]
```diff
@@ -62,11 +64,24 @@ async function start(): Promise<void> {
             faceCountMax: 4000,
         }
         const fabricKernel = new FabricKernel(engine, physics, dimensions)
+        const getFabric = (name: string) => {
```
Comment (mvanderh) [inline client/src/index.tsx]: Here, `name` is the construction code, right? If so, better to rename the parameter

File: client/src/storage/local-storage.ts [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    "2[2,2,2]",
+    "3[3,3,3[3,3,3]]L",
+    "2[2,2,2[2,2,2]]H",
+    "4[2=4]",
+    "6[0,1,1]",
+    "1[1,3[1,1,1],4]",
+    "3[3[3,3,3],3[3,3,3],3[3,3,3]]",
+    "3[3[3,3,3],3[3,3,3],3[3,3,3]]L",
+    "9[9,9,9]",
+    "2[1=2[3=2[2=2[3=2[1=2[3=2[2=2]]]]]]]X",
```
Comment (mvanderh) [inline client/src/storage/local-storage.ts]: Lookin' like a funky Regex

File: client/src/view/building-panel.tsx [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+    selection: ISelection,
+    setSelection: (s: ISelection) => void,
+}): JSX.Element {
+
+    function adjustment(up: boolean): number {
+        const factor = 1.1
+        return up ? factor : (1 / factor)
+    }
+
+    function SelectableButton({selectable, disabled}: { selectable: Selectable, disabled?: boolean }): JSX.Element {
```
Comment (mvanderh) [inline client/src/view/building-panel.tsx]: Ah that's cool, I didn't know you could use instance methods as JSX elements :)

Comment (mvanderh) [reply client/src/view/building-panel.tsx]: Oh wait, this is just a regular function, nevermind

File: client/src/view/building-panel.tsx [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+                    <MaybeSelectable selectable={Selectable.BAR}/>
+                    <MaybeSelectable selectable={Selectable.CABLE}/>
+                    <Cancel/>
+                </ButtonGroup>
+            </ButtonToolbar>
+        )
+    }
+
+    function Joint({joint}: { joint: IJoint }): JSX.Element {
+        const adjustUp = (up: boolean) => () => {
```
Comment (mvanderh) [inline client/src/view/building-panel.tsx]: `adjustUp` -> `adjust`

File: client/src/view/global-fabric-panel.tsx [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+import { useState } from "react"
+import {
+    FaBolt,
+    FaCompressArrowsAlt,
+    FaDownload,
+    FaParachuteBox,
+    FaRecycle,
+    FaRegFolder,
+    FaRegFolderOpen,
+    FaStarOfDavid,
```
Comment (mvanderh) [inline client/src/view/global-fabric-panel.tsx]: "Turn the fabric Jewish"

File: client/src/view/global-fabric-panel.tsx [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+                        <DropdownMenu>
+                            {loadFabricCode().map((code, index) => (
+                                <DropdownItem key={`Buffer${index}`} onClick={() => select(code, index)}>
+                                    {code}
+                                </DropdownItem>
+                            ))}
+                        </DropdownMenu>
+                    </ButtonDropdown>
+                </ButtonGroup>
+                &nbsp;&nbsp;&nbsp;
```
Comment (mvanderh) [inline client/src/view/global-fabric-panel.tsx]: Lol, might wanna use margins instead. Add a `className="mx-3"` to each ButtonGroup i reckon.

Comment (mvanderh) [reply client/src/view/global-fabric-panel.tsx]: "mx-3" standing for "60% (3/5) margin in the x direction"

File: client/src/view/tensegrity-view.tsx [hunk 13]
```diff
@@ -4,365 +4,77 @@
...
-            }
+        <div tabIndex={1} id="tensegrity-view" className="the-whole-page">
             <Canvas>
-                {!fabric ? undefined :
-                    <FabricView fabric={fabric}/>}
+                {!fabric ? undefined : <FabricView fabric={fabric} selection={selection} setSelection={setSelection}/>}
             </Canvas>
+            <GlobalFabricPanel
+                fabric={fabric}
+                constructFabric={code => {
```
Comment (mvanderh) [inline client/src/view/tensegrity-view.tsx]: Extract this anon func

File: client/src/fabric/tensegrity-brick.ts (hunk 5 above)
Comment (geralddejong) [reply client/src/fabric/tensegrity-brick.ts]: TRIANGLE_DEFINITIONS

File: client/src/view/global-fabric-panel.tsx (hunk 11 above)
Comment (geralddejong) [reply client/src/view/global-fabric-panel.tsx]: Haha.. looking and looking for a symbol that represents merging two triangles.  Well.. I found it!

File: client/src/view/global-fabric-panel.tsx (hunk 12 above)
Comment (geralddejong) [reply client/src/view/global-fabric-panel.tsx]: Ah lovely, i was looking for something like this.  Thx.

**Pull request merged**